permuta/perm.py

```diff
@@ -1247,39 +1103,34 @@ def count_non_inversions(self):
         n = len(self)
         return n * (n - 1) // 2 - self.count_inversions()
 
-    def non_inversions(self):
+    def non_inversions(self) -> Iterator[Tuple[int, int]]:
         """Yields the non_inversions of the permutation, i.e., the pairs i,j
         such that i < j and self[i] < self[j].
 
         Examples:
             >>> tuple(Perm((3, 0, 2, 1, 4)).non_inversions())
             ((0, 4), (1, 2), (1, 3), (1, 4), (2, 4), (3, 4))
         """
-        for i in range(len(self)):
-            for j in range(i + 1, len(self)):
-                if self[i] < self[j]:
-                    yield (i, j)
+        n = len(self)
+        for i, prev in enumerate(self):
+            for j in range(i + 1, n):
```

Comment:
This feels like the good place to use `islice`. Same for the inversions

Reply:
`islice` actually goes through the entire iterable and consumes it up to starting index so given `i=n-2`, the second loop would still iterate through the entire perm and just dump those up to `n-1`.

That's why I kept the second range.

Reply:
Ok make sense. I learned something